# 1. 回调地狱概念

所谓回调地狱（callback hell），即异步里面嵌套异步。异步任务无法保证顺序，如果我们想要保证顺序，就需要把后续的异步放在前一个异步的回调中，这样嵌套，就是回调地狱。

# 2. Promise

## 2.1 概述

Promise 通过链式调用来解决回调地狱，但是 Promise 一旦执行，无法中止。（即 Pendding 到 Rejected 或 Resolved 之间，无法停止。

Promise 本身不是异步的，其内部常常封装的异步。

通过 then 中返回 promise 实现链式调用，通过 catch 捕获错误（next 第二个参数也可以捕获错误）。

then 方法接收的值是上一次执行的结果。

```js
let promise = New Promise(function () {
  if (err)   {
      reject();
  }
  resolve();
})err
promise.then (promise())
	   .then (promise())
	   .then (function (data) {})
       .catch(function (err) {})
```

## 2.2 示例

【示例】promise 封装的 node.js 读取文件模块：（代码位对应 code 文件夹中的 “回调地狱-1” 文件夹）

readFsByPromise.js：

```js
let fs = require('fs');

function readFsByPromise(filePath) {
    return new Promise((res, rej) => {
        fs.readFile(filePath, 'utf-8', (err, data) => {
            if (err) {
                rej(err);
            }
            res(data);
        })
    })
}

let aFilePath = './a.txt';
let bFilePath = './b.txt';
let cFilePath = './c.txt';

readFsByPromise(aFilePath)
    .then((data) => {
        console.log(data)
        return readFsByPromise(bFilePath);
    })
    .then((data) => {
        console.log(data)
        return readFsByPromise(cFilePath);
    })
    .then((data) => {
        console.log(data)
    })
    .catch((err) => {
        throw err;
    })
```

## 2.3 使用场景

- 当业务需要多接口获取数据的时候，就可以通过 promise 来链式调用。（axios 就是使用的 promise）
- 使用 mongodb 数据库时，mongoose 的 API 支持 promise。

# 3. Generator

```js
function* gen () {
    yeid ...;
    yeid ...;
    yeid ...;
}
```

通过 next，依次调用 yeid 后面的方法，每次 next 的时执行的是上一代 yeid 到下一个 yeid 后面的内容。

且每一个 next 的参数表示的是上一次 yeid 表达式的返回值。第一个 next 前面没有 yeid 方法，所有第一个 next 方法的参数传入是无效的。

如果直接 return 方法，会返回给定的值并终止遍历。

# 4. async 和 await

这个是 Generator 的语法糖，详细可以看阮一峰老师的《ES6 标准入门》。

